### PR TITLE
Add opportunity to receive transaction with multiple addresses  

### LINKED ISSUE #1767: Pool Deposits broken
Pool deposits require the ability to accept multiple deposits from one TX id.

Version 1.8 and previous, had this ability for bitcoin based blockchains.

Version 1.9, fails when there is more than one deposit on a TX.  It only accepts the first one.     

For example, in Peatio 1.9, if each of the following addresses on this TX were exchange addresses, it would only credit a deposit to the first address:  (1JUNs2h2c5YN4ntqEadehrgwxQwrAtzfqh )
https://www.blockchain.com/btc/tx/4f44d9db153ff9345b62bb3e5bb0c91dc079fe4683a126262e8a80ff6f71a9e0

In Peaito 1.8, all of these addresses would be correctly credited.

Please restore this functionality, if possible.  For exchanges that don't want to accept pool deposits, a solution was already provided for them on issue #1744

### PR DESCRIPTION
**Pull request by mnaichuk**
closes #1767 
For BTC we will fetch number of sub transaction by `n` column from block `vout` and assigned to `txout` column in Deposit model. For erc20 we will fetch by `logIndex`column.

### CODE REVIEW

File: app/services/blockchain_service.rb [hunk 1]
```diff
@@ -51,10 +51,9 @@ def update_or_create_deposits!(deposits)
         # If deposit doesn't exist create it.
         deposit = Deposits::Coin
                     .where(currency: currencies)
-                    .find_or_create_by!(deposit_hash.slice(:txid)) do |deposit|
+                    .find_or_create_by!(deposit_hash.slice(:txid, :address)) do |deposit|
```
Comment (ysv) [inline app/services/blockchain_service.rb]: we need to find_or_create_by txid and vout

File: app/models/deposits/coin.rb [hunk 2]
```diff
@@ -7,7 +7,6 @@ class Coin < Deposit
 
     validate { errors.add(:currency, :invalid) if currency && !currency.coin? }
     validates :address, :txid, :txout, presence: true
-    validates :txid, uniqueness: { scope: %i[currency_id txout] }
```
Comment (ysv) [inline app/models/deposits/coin.rb]: This validation is still ok

File: app/services/blockchain_service/bitcoin.rb [hunk 3]
```diff
@@ -46,12 +46,12 @@ def build_deposits(block_json, block_id)
         .fetch('tx')
         .each_with_object([]) do |tx, deposits|
 
-        payment_addresses_where(address: client.to_address(tx)) do |payment_address|
+        payment_addresses_where(address: client.to_address(tx)).each_with_index do |payment_address, i|
```
Comment (ysv) [inline app/services/blockchain_service/bitcoin.rb]: I'm not sure if we need this changes
I reckon that the problem is in https://github.com/rubykube/peatio/pull/1792/files#diff-12ec28c62ee4ee7ef72503aae2a92392R54

**Review by ysv — CHANGES_REQUESTED** [review]

**Comment by ysv** [comment]
@mnaichuk Add specs for this context
1. Multiple vout with platform addresses
2. Multiple vout with same platform address (if this is possible)

File: lib/blockchain_client/ethereum.rb [hunk 4]
```diff
@@ -41,11 +41,11 @@ def from_address(tx)
       normalize_address(tx['from'])
     end
 
-    def build_transaction(txn, current_block_json, currency)
+    def build_transaction(txn, current_block_json, payment_address, currency)
```
Comment (mnaichuk) [inline lib/blockchain_client/ethereum.rb]: `address`

File: lib/blockchain_client/ripple.rb [hunk 5]
```diff
@@ -31,7 +31,8 @@ def build_transaction(tx:, currency:)
     def build_entries(tx, currency)
       [
         {
-          amount:  convert_from_base_unit(tx.fetch('Amount'), currency)
+          amount:  convert_from_base_unit(tx.fetch('Amount'), currency),
+          txout:   0
```
Comment (mnaichuk) [inline lib/blockchain_client/ripple.rb]: `nil`

File: spec/services/blockchain_service/ethereum_spec.rb [hunk 6]
```diff
@@ -183,6 +187,68 @@ def request_body(block_number, index)
...
+      # File with real json rpc data for bunch of blocks.
+      let(:block_file_name) { '2621839-2621843.json' }
+
+      # Use rinkeby.etherscan.io to fetch transactions data.
+      let(:expected_deposits) do
+        [
+          {
+            amount:   '0x1e8480'.hex.to_d / currency.base_factor,
+            address:  '0xe3cb6897d83691a8eb8458140a1941ce1d6e6daa',
+            txid:     '0x826555325cec51c4d39b327e563ce3e8ee87e27be5911383f528724a62f0da5d'
```
Comment (mnaichuk) [inline spec/services/blockchain_service/ethereum_spec.rb:202]: `vout`

File: spec/services/blockchain_service/ethereum_spec.rb [hunk 7]
```diff
@@ -183,6 +187,68 @@ def request_body(block_number, index)
...
+          },
+          {
+            amount:   '0x1e8480'.hex.to_d / currency.base_factor,
+            address:  '0xe3cb6897d83691a8eb8458140a1941ce1d6e6dab',
```
Comment (mnaichuk) [inline spec/services/blockchain_service/ethereum_spec.rb]: use another address

**Review by ysv — APPROVED** [review]

**Comment by mod** [comment]
@pkucherenk0 do you confirm ?

**Comment by pkucherenk0** [comment]
Can confirm that BTC deposit to multiple addresses works well, but I haven't done ERC20 deposit to multiple addresses, yet.

**Comment by Fair-Exchange** [comment]
Thank you so much, will try this tonight

**Pull request merged**